src/base.jl

```diff
@@ -1 +1,34 @@
+
 @functor Base.RefValue
+
+functor(::Type{<:Base.ComposedFunction}, x) = (outer = x.outer, inner = x.inner), y -> Base.ComposedFunction(y.outer, y.inner)
+
+using LinearAlgebra
+# The reason for these is to let W and W' be seen as tied weights in Flux models.
+# Can't treat ReshapedArray very well, as its type doesn't include enough details for reconstruction.
+
+functor(::Type{<:Adjoint}, x) = (parent = _adjoint(x),), y -> adjoint(only(y))
```

Comment:
Yes. The point of `_adjoint` is the it digests both, it will `adjoint(dx::Matrix)` or unwrap `dx::Adjoint` (both natural), and deal with `dx::NamedTuple`. I'm not honestly 100% sure why the structural ones show up, there is some chance that could always be avoided.

But the fact that `functor` can be used on `dx` with a type from `x` is, I now think, the key feature to make it useful for Optimisers.